### ISSUE TITLE
Reduce memory usage within VS

### ISSUE THREAD

**Issue by jamesmanning** [icsharpcode/CodeConverter#605]
### Steps to reproduce
want to make a simpler repro but can't seem to find any log output from the conversion - does it write intermediate results to any log locations that could help isolate the cause? It's a large source project and I'd much rather share a far more minimal repro.

### Error message shown

```
---------------------------
Code converter 8.1.6 - Visual Studio 16.0.30329.88
---------------------------
An error has occured during conversion - press Ctrl+C to copy the details: System.InvalidOperationException: Exception of type 'System.InvalidOperationException' was thrown.
   at System.Runtime.CompilerServices.ConditionalWeakTable`2.VerifyIntegrity()
   at System.Runtime.CompilerServices.ConditionalWeakTable`2.TryGetValue(TKey key, TValue& value)
   at Microsoft.CodeAnalysis.GreenNode.GetAnnotations()
   at Microsoft.CodeAnalysis.GreenNode.GetAnnotations(String annotationKind)
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.GetErrorsFromAnnotations(SyntaxNode convertedNode)
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<SecondPassLoggedAsync>d__22.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at ICSharpCode.CodeConverter.Shared.AsyncEnumerableTaskExtensions.<ParallelSelectAwait>d__1`2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.GetResult(Int16 token)
   at ICSharpCode.CodeConverter.Shared.AsyncEnumerableTaskExtensions.<ParallelSelectAwait>d__1`2.System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)
   at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.<MoveNextCore>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Linq.AsyncIteratorBase`1.<MoveNextAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Linq.AsyncIteratorBase`1.<MoveNextAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<Convert>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<Convert>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.GetResult(Int16 token)
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<Convert>d__19.System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<ConvertProjectContents>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<ConvertProjectContents>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.GetResult(Int16 token)
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<ConvertProjectContents>d__18.System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<WithProjectFile>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<WithProjectFile>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.GetResult(Int16 token)
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<WithProjectFile>d__13.System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<ConvertProject>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<ConvertProject>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.GetResult(Int16 token)
   at ICSharpCode.CodeConverter.Shared.ProjectConversion.<ConvertProject>d__12.System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Linq.AsyncEnumerable.SelectManyAsyncIterator`2.<MoveNextCore>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Linq.AsyncIteratorBase`1.<MoveNextAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Linq.AsyncIteratorBase`1.<MoveNextAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Linq.AsyncEnumerable.ConcatAsyncIterator`1.<MoveNextCore>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Linq.AsyncIteratorBase`1.<MoveNextAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Linq.AsyncIteratorBase`1.<MoveNextAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at ICSharpCode.CodeConverter.VsExtension.CodeConversion.<ConvertProjectUnhandled>d__24`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at ICSharpCode.CodeConverter.VsExtension.CodeConversion.<ConvertProjectUnhandled>d__24`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.GetResult(Int16 token)
   at ICSharpCode.CodeConverter.VsExtension.CodeConversion.<ConvertProjectUnhandled>d__24`1.System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)
   at ICSharpCode.CodeConverter.VsExtension.CodeConversion.<WriteConvertedFilesAndShowSummaryAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at ICSharpCode.CodeConverter.VsExtension.CodeConversion.<WriteConvertedFilesAndShowSummaryAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ICSharpCode.CodeConverter.VsExtension.CodeConversion.<>c__DisplayClass13_0`1.<<ConvertProjectsAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask.<JoinAsync>d__68.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ICSharpCode.CodeConverter.VsExtension.CodeConversion.<ConvertProjectsAsync>d__13`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ICSharpCode.CodeConverter.VsExtension.ConvertVBToCSCommand.<SolutionOrProjectMenuItemCallbackAsync>d__23.MoveNext()
---------------------------
OK   
---------------------------

```
### Details
Not sure why the error dialog gives VS version of 16.0.30329.88, the help -> about shows it's version 16.7.1

My next step will be to attempt to run the extension under the debugger, but since the process took around 20 minutes normally, I'm not sure how well or how long it'll take like that.

If there's any additional log I can find/share or enable that would help provide more context, please let me know. :)

**Comment by GrahamTheCoder** [comment]
Thanks for the report. It's a very unusual one - a BCL method called VerifyIntegrity is throwing. All the converter was trying to do was asking roslyn to look up a string in a dictionary really. I guess since it's a weak reference then if your system is seriously short of ram that could cause an issue maybe.

The first thing to check is whether the error occurs on a brand new empty project. If that's broken too, run a repair install on VS.

Otherwise, see whether it can be reproduced by running the same conversion that originally caused it.

If so, the "code converter" pane of the output window should be displaying a log of progress. You may see several files say "started" don't have a corresponding "finished" entry. 
If you individually convert those files you may be able to narrow down the file and then a particular snippet.

Running in a debugger as you suggest would also likely let you see a filename on the call stack at the point of the error.

**Comment by jamesmanning** [comment]
@GrahamTheCoder thanks for the fast response!  I really appreciate the effort you've been making on the converter.

While the machine it was running on had enough memory (8GB and nothing else significant running AFAIK), I think you're right about the memory situation being a cause.  I see now that there's a 3.5GB dump file created at the time in temp of **VsLowMemoryDump**ccf94c14-2f60-49c8-8988-0e01190b1213.dmp so in its 32-bit process I guess it ran out of available memory and I could definitely see any weak references being gone.

I tried on another machine with 16GB and it had a different behavior - VS crashed when trying to run the conversion - event viewer of application log shows VS crashing during the conversion after about 9 minutes,

```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.OutOfMemoryException
   at System.Threading.ThreadPoolWorkQueue.EnsureCurrentThreadHasQueue()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

However, running codeconv from the command-line worked fine so given the size of this project (which I assume is the underlying cause) is a likely outlier, this is probably not worth fixing since I'm guessing it'd mean trying to start up a 64-bit process (codeconv.eve or otherwise) and doing the communication back to the VS extension, and that complexity is likely not worth it.

At best it'd be nice to have some kind of error dialog that caught this situation and told the user to just use the command-line instead, but given the OOM exception and devenv crashing, I'm not sure how feasible that would be anyway.

Given there's no bug in codeconv here, it's just a victim of the VS 32-bit process running up against memory limits, and since the user has a perfectly good workaround in the command-line, I'm going to go ahead and close this since I don't think there's anything to fix.

**Comment by GrahamTheCoder** [comment]
Thanks for getting back to me with the extra info. Yeah if VS crashes with an OOM there's little that can be done at that point. But maybe I can make it less likely.
In the VS conversion, did you get any output files at all, or was it still in "phase 1" (if you saw the output window)?
Could you give me a sense of how big the project was that was being converted? e.g. file count and/or lines of code count if you have it?

I'll have a think about ways of limiting "work in progress" e.g.:
* Investigate if there are any references held to objects longer than they're needed (e.g. the source project)
* Usually it converts N files in parallel for an N core computer, it's possible that backing off from this would ease the memory pressure - could make it a user option.

Do raise any issues for things you see in the command line conversion result too. It's actually relatively new and I've had almost no feedback on it so far, so even hearing it worked is a great start, thanks!

**Comment by GrahamTheCoder** [comment]
Updating to triage so I don't lose track of it

**Comment by jamesmanning** [comment]
@GrahamTheCoder the failed ones definitely wrote files - the stack shows it was in SecondPassLoggedAsync

The stats on the VB project being converted: ~25k lines across 25 files, with the 'main' file having a little over 10k lines in its *.designer.vb file and a little over 7k lines in its *.vb / codebehind file (yes, it's horrific like that).  Comparing the results of the crashed-VS attempt and the command-line result, that one huge class was the only C# file not written (a Resources.resx and the csproj was also not written, but I'm assuming the large class was the problem causing OOM)

AFAICT the extension conversion process blocks the UI thread (I tried copying the "Code Converter" output pane contents with no luck, and the VS interface changed with that the grayed-out / not-responding look, so changing over to invoking the command-line (and just taking the output lines from the OutputDataReceived and maybe ErrorDataReceived and writing them to the output pane as it progresses) would unblock the UI thread, remove the entire OOM of the devenv process, and result in more coverage/testing of the command-line since the extension would invoke it. 😄 

Just a thought.  The actual resulting code has its own issues (I'm guessing a significant number just based on an initial attempted build) that I need to file so thought you might be able to solve the more-command-line-testing and OOM-on-large-project issues at the same time in a hopefully relatively simple way.

The difference also showed that the extension attempt wrote byte-order marks at the start of the output C# files and the command-line version didn't but I don't think that really matters, just figured I'd mention it since it showed up in the diff.

I'll work on the build issues with the resulting C# code and filing issues for them, trying to create minimal repros for each. I vaguely remember from a previous attempt that the issues were mainly are VB-isms that probably just don't have an equivalent - I think one was a missing/implied function for VB winforms projects and another was use of indexed properties, but I'll file them separately. 😄

**Comment by GrahamTheCoder** [comment]
Interesting. None of the conversion process should be able to use the ui thread. Though the converter aims to use a thread per core and so can somewhat starve the ui thread, I haven't seen that happen in general. Obviously while doing the OOM crash dump the ui will be locked up though.

One of the test cases I use is  a 250k line project, so I suspect it's the few individual large files causing the memory issues in roslyn.

Being in the vs process allows using the exact workspace the user normally builds in, which is handy. But at some point perhaps an out of process call will be necessary, I'll investigate other options first though.

For the command line did it convert immediately or did you have to specify the  -besteffort flag?

**Comment by jamesmanning** [comment]
The command line converted immediately, I only passed the solution file path, no other parameters. Worked great!

**Comment by GrahamTheCoder** [comment]
The latest version of the extension is now compatible with [VS 2022 preview](https://visualstudio.microsoft.com/vs/preview/vs2022/), which is 64-bit and hence unlikely to experience this issue except on huge solutions/files, or machines with a lack of RAM. I'll close the issue until someone sees such a problem
